contracts/contracts/token/OUSD.sol

```diff
@@ -511,93 +493,179 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
      * address's balance will be part of rebases and the account will be exposed
      * to upside and downside.
      */
-    function rebaseOptIn() public nonReentrant {
+    function rebaseOptIn() external {
         _rebaseOptIn(msg.sender);
     }
 
     function _rebaseOptIn(address _account) internal {
-        require(_isNonRebasingAccount(_account), "Account has not opted out");
-
-        // Convert balance into the same amount at the current exchange rate
-        uint256 newCreditBalance = _creditBalances[_account]
-            .mul(_rebasingCreditsPerToken)
-            .div(_creditsPerToken(_account));
-
-        // Decreasing non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.sub(balanceOf(_account));
+        // prettier-ignore
+        require(
+            alternativeCreditsPerToken[_account] > 0 ||
+                // new empty contracts that haven't been yet autoMigrated to StdNonRebasing can
+                // explicitly call `rebaseOptIn`. Side effect is that also already rebasing EOA
+                // accounts that have 0 balance can call RebaseOptIn
+                balanceOf(_account) == 0
+            ,
+            "Account must be non-rebasing"
+        );
+        RebaseOptions state = rebaseState[_account];
+        // prettier-ignore
+        require(
+            state == RebaseOptions.StdNonRebasing ||
+                state == RebaseOptions.NotSet,
+            "Only standard non-rebasing accounts can opt in"
+        );
 
-        _creditBalances[_account] = newCreditBalance;
+        uint256 balance = balanceOf(_account);
```

Comment:
This should move up to the start of the function, and then the check on zero balance should use the now already loaded `balance` variable.

Reply:
nice catch: https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/dc803f2518ac3e5f06cb5eb040b875f90b4425bb